File: src/modules/cmdpal/ext/SamplePagesExtension/OnLoadPage.cs
// Copyright (c) Microsoft Corporation
// The Microsoft Corporation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System;
using System.Collections.Generic;
using System.Globalization;
using Microsoft.CommandPalette.Extensions;
using Microsoft.CommandPalette.Extensions.Toolkit;
using Windows.Foundation;

namespace SamplePagesExtension;

internal sealed partial class OnLoadPage : IListPage
{
    private readonly List<ListItem> _items = new();

    public IIconInfo Icon => new IconInfo("\uE8AB"); // switch

    public string Title => "Load/Unload sample";

    public string PlaceholderText => "This page changes each time you load it";

    public ICommandItem EmptyContent => new CommandItem() { Icon = new IconInfo("\uE8AB"), Title = "This page starts empty", Subtitle = "but go back and open it again" };

    public IFilters Filters => null;

    public IGridProperties GridProperties => null;

    public bool HasMoreItems => false;

    public string SearchText => string.Empty;

    public bool ShowDetails => false;

    public OptionalColor AccentColor => default;

    public bool IsLoading => false;

    public string Id => string.Empty;

    public string Name => "Open";

#pragma warning disable CS0067 // The event is never used
    public event TypedEventHandler<object, IPropChangedEventArgs> PropChanged;

    private event TypedEventHandler<object, IItemsChangedEventArgs> InternalItemsChanged;
#pragma warning restore CS0067 // The event is never used

    public event TypedEventHandler<object, IItemsChangedEventArgs> ItemsChanged
    {
        add
        {
            InternalItemsChanged += value;
            var nowString = DateTime.Now.ToString("T", CultureInfo.CurrentCulture);
            var item = new ListItem(new NoOpCommand())
            {
                Title = $"Loaded {nowString}",
                Icon = new IconInfo("\uECCB"), // Radio button on
            };
            _items.Add(item);
        }

        remove
        {
            InternalItemsChanged -= value;
            var nowString = DateTime.Now.ToString("T", CultureInfo.CurrentCulture);
            var item = new ListItem(new NoOpCommand())
            {
                Title = $"Unloaded {nowString}",
                Icon = new IconInfo("\uECCA"), // Radio button off
            };
            _items.Add(item);
        }
    }

    public IListItem[] GetItems() => _items.ToArray();

    public void LoadMore()
    {
    }
}

File: src/modules/cmdpal/ext/SampleJSExtension/src/pages/onLoadPage.ts
// Copyright (c) Microsoft Corporation
// The Microsoft Corporation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

import { ListItemBase, ListPageBase, NoOpCommand } from '@microsoft/cmdpal-sdk';
import type { IListItem } from '@microsoft/cmdpal-sdk';
import { glyphIcon } from '../util.js';

/**
 * A page that adds one entry each time it opens. It mirrors the load side of the
 * C# `OnLoadPage`.
 *
 * The JS protocol has no explicit page load event yet, so the sample appends a
 * "Loaded" entry when the host asks for items during page open.
 */
export class OnLoadPage extends ListPageBase {
  readonly id = 'on-load-page';
  readonly name = 'Open';
  readonly title = 'OnLoad sample';

  override icon = glyphIcon('\uE8AB');
  override placeholderText = 'This page changes each time you load it';
  override emptyContent = new ListItemBase({
    command: new NoOpCommand('on-load-empty'),
    title: 'This page starts empty',
    subtitle: 'but go back and open it again',
    icon: glyphIcon('\uE8AB'),
  });

  private readonly items: IListItem[] = [];

  override getItems(): IListItem[] {
    const now = new Date().toLocaleTimeString();
    this.items.push(
      new ListItemBase({
        command: new NoOpCommand(`on-load-${this.items.length}`),
        title: `Loaded ${now}`,
        icon: glyphIcon('\uECCB'),
      }),
    );
    return [...this.items];
  }
}
